ZenPacks/zenoss/ZenPackLib/lib/helpers/utils.py

```diff
@@ -0,0 +1,248 @@
+    removed = d2_keys - d1_keys
+    modified = {o : (d1[o], d2[o]) for o in intersect_keys if d1[o] != d2[o]}
+    same = set(o for o in intersect_keys if d1[o] == d2[o])
+    LOG.warn('ADDED: {}'.format(added))
+    LOG.warn('REMOVED: {}'.format(removed))
+    LOG.warn('MODIFIED: {}'.format(modified))
+
+
+def get_optimized_yaml(data):
+    """
+        return optimized YAML representing ZenPackSpec
+    """
+    # set DEFAULTS throughout
+    descend_defaults(data)
+    # sort
+    ordered = sort_yaml_data(data)
+    # optimized output
+    return yaml.dump(ordered,  default_flow_style=False, Dumper=Dumper).replace('!!map','')
+
+
+def descend_defaults(input):
+    """
+        descend YAML dictionary and optimize DEFAULTS
+    """
+    for k, v in input.items():
+        if not isinstance(v, dict):
+            continue
+        if k == 'DEFAULTS': 
+            continue
+        if k in ['classes', 'properties', 'thresholds', 'datasources', 'datapoints', 'graphs', 'relationships','graphpoints']:
+            if len(v.keys()) > 1:
+                set_defaults(v)
+        descend_defaults(v)
+
+
+def set_defaults(input):
+    """
+        if dict has multiple entries, look for identical key/value 
+        pairs and move them to DEFAULTS
+    """
+    defaults = {}
+    # get list of potential defaults
+    for name, params in input.items():
+        if not isinstance(params, dict): 
+            continue
+        for k,v in params.items():
+            if isinstance(v, dict): 
+                continue
+            if k not in defaults.keys():
+               defaults[k] = v
+    # disqualify defaults if not universal
+    for k, v in defaults.items():
+        params = input.get(k)
+        for name, params in input.items():
+            if not isinstance(params, dict):
+                continue
+            value = params.get(k)
+            if v != value:
+                if k in defaults.keys():
+                    defaults.pop(k)
+    if len(defaults.keys()) == 0:
+        return
+    # remove universal parameters from individual spec
+    for name, params in input.items():
+        if not isinstance(params, dict): 
+            continue
+        for k in params.keys():
+            if k in defaults.keys():
+                params.pop(k)
+    # add defaults to original data
+    if 'DEFAULTS' in input.keys():
+        input['DEFAULTS'].update(defaults)
+    else:
+        input['DEFAULTS'] = defaults
+
+
+def sort_yaml_data(data):
+    """
+        Sort YAML data before outputting
+    """
+    if not isinstance(data, dict):
+        return data
+    ordered = OrderedDict()
+    fields = ['zProperties', 'class_relationships', 'classes', 'device_classes', 'DEFAULTS',
+              'base','label','order', 'properties', 'monitoring_templates', 
+              'dynamicview_views', 'dynamicview_relations',
+              'impacts', 'impacted_by', 
+              'relationships']
+    if 'name' in data.keys():
```

Comment:
this could be moved within the for loop below.  also, you're calling data.keys() 3 times.  You can make a variable out of it, `data_keys = data.keys()`